crates/ra_ide/src/syntax_highlighting.rs

```diff
@@ -592,6 +613,52 @@ fn highlight_element(
                         HighlightTag::SelfKeyword.into()
                     }
                 }
+                T![ref] => {
+                    let modifier: Option<HighlightModifier> = (|| {
+                        let bind_pat = element.parent().and_then(ast::BindPat::cast)?;
+                        let parent = bind_pat.syntax().parent()?;
+
+                        let ty = if let Some(pat_list) =
+                            ast::RecordFieldPatList::cast(parent.clone())
+                        {
+                            let record_pat =
+                                pat_list.syntax().parent().and_then(ast::RecordPat::cast)?;
+                            sema.type_of_pat(&ast::Pat::RecordPat(record_pat))
+                        } else if let Some(let_stmt) = ast::LetStmt::cast(parent.clone()) {
+                            let field_expr =
+                                if let ast::Expr::FieldExpr(field_expr) = let_stmt.initializer()? {
+                                    field_expr
+                                } else {
+                                    return None;
+                                };
+
+                            sema.type_of_expr(&field_expr.expr()?)
+                        } else if let Some(record_field_pat) = ast::RecordFieldPat::cast(parent) {
+                            let record_pat = record_field_pat
+                                .syntax()
+                                .parent()
+                                .and_then(ast::RecordFieldPatList::cast)?
+                                .syntax()
+                                .parent()
+                                .and_then(ast::RecordPat::cast)?;
+                            sema.type_of_pat(&ast::Pat::RecordPat(record_pat))
+                        } else {
+                            None
+                        }?;
+
+                        if !ty.is_packed(db) {
+                            return None;
+                        }
+
+                        Some(HighlightModifier::Unsafe)
+                    })();
+
+                    if let Some(modifier) = modifier {
+                        h | modifier
+                    } else {
+                        h
+                    }
+                }
```

Comment:
This is an uncomfrotable amount of logic for syntax highlighting. Let's add an `Semancits::is_unsafe_expr(expr: &ast::Expr) -> bool` method.

Reply:
I'm a bit confused on how that method signature could be used to reduce the code above. Only the `LetStmt` branch above references an `expr`. Would it make sense for something like `Semantics::is_ref_unsafe(SyntaxNode<RustLanguage>) -> bool`?

Reply:
It’s better stick to strongly typed api. If both patterns and exprs can be unsafe, that we should have _pat and _expr method. The goal is not to reduce the code, the goal is to not put intricate knowledge of language semantics into syntax highlighter.

Reply:
I kept it all in one `_pat` method, since `ref` relies entirely patterns to check whether it's unsafe or not. The only expr usage is when checking the initializer of the `let` statement. It seems easier to follow now, and I added comments to hopefully clarify what would otherwise be a pretty confusing function.